File: Leetcode/234.cpp
// https://leetcode.com/problems/palindrome-linked-list/
// Solution 1
// C++

#include <iostream>
#include <stack>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
    public :
        bool isPalindrome(ListNode* head) {
            // stack
            stack <int> s;
            ListNode* node = head;    

            // push a node into the stack 
            while (node != NULL) {
                s.push(node->val); 
                node = node->next; // move on to the next node
            }

            // traverse
            node = head;
            while(node != NULL) {
                int top = s.top(); // get the element in the top of the stack
                s.pop();

                // compare the popped element with the current node's data
                if(top != node->val)
                    return false;
                node = node->next; // move on to the next node
            }
            return true;
        }
};

int main() {
    Solution s1;

    ListNode *head = new ListNode(1);
    head->next = new ListNode(4);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(2);
    head->next->next->next->next = new ListNode(1);

    if(s1.isPalindrome(head)) {
        cout << "Linked List is a palindrome."<<"\n";
    }
    else {
        cout << "Linked List is not a palindrome."<<"\n";
    }
    return 0;
}

File: Leetcode/234.py
# https://leetcode.com/problems/palindrome-linked-list/
# Solution 2
# Python

from collections import deque

class ListNode :
    def __init__(self, val = 0, next = None) :
        self.val = val
        self.next = next
    
class Solution :
    def isPalindrome(self, head : ListNode) -> bool :
        dq = deque()

        # insert a node to the right end of the deque
        while head != None :
            dq.append(head.val)
            head = head.next
        while len(dq) > 1 :
            right = dq.pop() # pop an element from the right end of the deque
            left = dq.popleft() # pop an element from the left end of the deque
            if right != left : # compare the values
                return False
        return True

if __name__ == '__main__' :

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(2)
    head.next.next.next.next = ListNode(1)

    if Solution.isPalindrome(Solution, head) :
        print('Linked List is a palindrome.')
    else :
        print('Linked List is not a palindrome.')